lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -265,6 +268,52 @@ struct CircuitLoweringState {
     binds.push_back(op);
   }
 
+  /// For a given Type Alias, return the corresponding AliasType. Create and
+  /// record the AliasType, if it doesn't exist.
+  hw::TypeAliasType getTypeAlias(Type rawType, BaseTypeAliasType firAliasType,
+                                 Location typeLoc, StringRef suffix) {
+
+    auto typeName = firAliasType.getName();
+    assert(typeName && "getTypeAlias can only be called for a named Type");
+    // If the lowered type already exists, return that.
+    auto iterT1 = firrtlTypeToAliasTypeMap.find(firAliasType);
+    if (iterT1 != firrtlTypeToAliasTypeMap.end())
+      return iterT1->second;
+    // Mutex, to update the top level Module, and add the TypeDeclOp inside the
+    // TypeDeclOp.
+    std::lock_guard<std::mutex> lock(typeScopeMutex);
+    // The typescope is created only at the top level module.
+    Block &topLevelBlock = circuitOp->getParentRegion()->getBlocks().back();
```

Comment:
So, the ports are not lowered in parallel, so name conflict with ports across modules should be fine. Types within a module are lowered sequentially, that should also be fine. I was thinking type name conflict across modules can be resolved by appending the module name to the type. But, you are right, the current implementation only appends the module name on conflict, which can result in non-deterministic result. Did we specify this type name conflict behavior in the spec ?

Reply:
I agree that type alias which appears in ports are fine but I think type aliases in modules are lowered parallely so it looks unsafe to mutate single type decl. 
I guess currently type name conflicts would not happen because type aliases are unique at parser level, and type aliases will be dropped when inner types are mutated.

Reply:
> So, here I used the lock on line 284, that should be make mutating the type decl safe, right ?

Yes, I think this is thread safe but operations order would be non-deterministic.